### DISCUSSION TITLE
Better ordering of groups of globals & collections

### DISCUSSION THREAD

**Discussion by echocrow** [payloadcms/payload#1277] [Feature Requests & Ideas]
Really appreciate the recently added feature to set a custom `group` values on collections and globals! This alone already allows organizing collections & globals into more meaningful & user-friendly groups.

The last piece missing for us to make the overall organization perfect is the sorting order across collections & globals. It looks like the order of groups is currently defined by the order in which collections & globals as passed into the payload config. Where this falls short is if we want e.g. a group of globals at the top.

A practical example:
- Layout
  - Header _(global)_
  - Footer _(global)_
  - Homepage _(global)_
- Catalogs
  - Catalog Type A _(collection)_
  - Catalog Type B _(collection)_
- Content
  - Some Other Global _(global)_
  - Some Other Collection _(collection)_
- …

This currently does not seem supported, outside of creating a custom navbar component just for this.

Some potential approaches how this could be achieved:
- A new global payload config entry, e.g. `admin.groups` or `admin.knownGroups` of type `string[]`
  - this would specify the primary order of groups; if a group ends up with no collections, it can be skipped.
  - if collections are listed with a group not in the list, they get appended
  - might be the easiest add, but it does not allow for granular intra-group ordering (e.g. a group with a global, a collection, and then another global, in this order)
- A new entry on collections & globals, e.g. `admin.sort` or `admin.order` of type `number`
  - would allow for perfect, granular control over sorting orders across all types
  - would also be a bit ugly to maintain, as weights are spread across collections & globals
- Migrating from separate `collections` vs `globals` arrays on the payload config into a single  entry, e.g. `schemas`
  - this array would then define the order of all collections/entries in a single place
  - does break the current schema, but could be done in a backwards-compatible way
  - would require some extra `global` flag on collections vs globals to differentiate the two
  - would definitely muddy the water for the payload config while backwards compatibility is maintained

**Comment by echocrow** [comment]
Happy to help out with this btw! But would like some guidance on the approach first.

Assuming no refactors along the lines of merging collections and globals are planned, I reckon a simple `admin.groups` `string[]` property on the config might be a simple, incremental feature, and potentially a good first issue?

**Reply by DanRibbens** [reply]
@echocrow I remember thinking about this exact limitation being possible when writing the groups feature and dismissing it in my head as an unlikely edge case. Haha! This discussion appeared much quicker than I expected.

How about this?

Let's add a new property at the top of the config in `admin.sortGroups`. This will be a function that is called with the default order that Payload currently does but allows you to return your desired order instead.

```ts
buildConfig({
  admin: {
	sortGroups: (groups: string[]]): string[] => {
		// you could write a function to sort and return the or hard-code your desired order
		return ['Layout', 'Catalogs', 'Content' //...];
    }
  },
  collections: [ //...],
  globals: [//...]
});
```

I'd consider not showing the groups if they don't come back from the function, but this would be a strange workaround for hiding collections, curious what others think about this. Maybe allow `group: false` to be set on the collections or globals to hide them from the nav as another feature that addresses it more direct?

This way we're not introducing any deprecation changes, the global and collection config types are not complicated and there is not an arbitrary sort number to maintain. For the 99% of people already using this feature they can continue to rely on Payload's default way of handling it. Then the others that have your issue (global only group coming first) has a way to override it however way they want in code.

I'm open to feedback on how to imporove this idea. Full transparency, this is not a high priority for our team at the moment.

**Reply by echocrow** [reply]
> Haha! This discussion appeared much quicker than I expected.

guilty as charged 😅

> I'd consider not showing the groups if they don't come back from the function, but this would be a strange workaround for hiding collections, curious what others think about this. Maybe allow group: false to be set on the collections or globals to hide them from the nav as another feature that addresses it more direct?

agreed that this seems like a strange workaround. a more explicit `group: false` or `hidden: true` on the respective collection would seem more declarative

> Let's add a new property at the top of the config in admin.sortGroups. This will be a function that is called with the default order that Payload currently does but allows you to return your desired order instead.

this works, but might be a bit more verbose than necessary? especially if, at the end, groups _not_ returned or explicitly listed should still be listed.

are there other benefits to making this a function, rather than a `string[]`? e.g.
```js
buildConfig({
  collections: [ //...],
  globals: [//...],
  admin: {
    groups: ['Layout', 'Catalogs', 'Content', /*...*/],
  },
});
```

that said, a function could provide the exact same, and potentially offer more flexibility around sorting if, for whatever reason, necessary?
personally a fan of keeping things simple though; support for a function could also be added later, while still supporting a string array too (i.e. expanding this config to `string[] | (groups: string[]) => string[]`)

with the `string[]` approach (or maybe even the function), the following steps could handle the normal cases and the edge cases of "missing"/unlisted groups:
1. use `config.admin.groups` as starting ground (default to empty array)
2. for each collection:
    1. get the group (default to `Collections` - or whatever the current collection default is)
    2. append to groups list if not already present
3. for each global:
    1. get the group (default to `Globals` - or whatever the current global default is)
    2. append to groups list if not already present
4. remove empty groups (probably no reason for them to stick around?)

> Full transparency, this is not a high priority for our team at the moment.

yeah I assumed as much, given this is probably a niche case. that said, i'd be happy to work on a PR if there is a change this feature may get accepted

**Reply by DanRibbens** [reply]
Yeah I guess I don't feel too strongly about function vs string[] or both.

We can just concat any missed groups on the end, whether it is a function or not. That would be super simple.

Agreed, `admin.hidden: boolean` makes the most sense on hidding collections from the admin UI.

I'd like to get consensus on the feature from other devs too before we consider before starting on a PR.

Thanks, stay tuned!

**Comment by ChrisGV04** [comment]
This would be really useful to order the collections and globals in a meaningful way. It's kind of odd in some cases where we can't order the items in a group to match the logical way we would expect to find them in.

Is there any news on this topic?

**Comment by esonesson** [comment]
Anyone else found a solution for this? I want to be able to order and mix globals and collections within a group specifically.

For example, I want to have the start page as a singleton/global above the collection pages in the group content

**Comment by tori-takashi** [comment]
Is there any news on this topic?

**Comment by ClemannD** [comment]
Found a very hacky _partial_ workaround.

I created a dummy collection like 
```
import type { CollectionConfig } from 'payload';

export const PageNavPlaceholder: CollectionConfig = {
  slug: ' ',
  admin: {
    group: 'Pages',
  },
  fields: [],
};
```

Then in the config:
```

  globals: [
    HomePage,
    AboutPage,
    ContactPage,
    GalleryPage,
    ...
  ],
  collections: [Users, PageNavPlaceholder, ...],
```

since the collections order dictates what the order of the groups are, using this placeholder I can get my Pages group to appear before the others. 

--- 

Of course this is hacky and it would be great if this was a supported feature

**Comment by Mionja** [comment]
Is there any update on this?

**Comment by julio-benavente** [comment]
Maybe not the solution we're all waiting for, but the navbar can be customized with a little bit of CSS.

```
.nav__wrap .nav-group {
    order: 90;
}

#nav-group-Pages {
    order: 1;
}

#nav-group-Globals {
    order: 2;
}

.nav__wrap .nav__controls {
    order: 999;
}
```

**Reply by acerspyro** [reply]
Careful with the CSS order property, it's probably one of the worst implemented CSS properties IMO.

Since CSS only controls styling, screen readers only look at the DOM ordering. This means that screen readers will read the groups out of order, which is bad if you depend on a screen reader but still have some sight as it can be confusing.

I will personally be going with a JS approach to re-order the DOM elements.

**Comment by devinoldenburg** [comment]
Opened #17097 to track this as a feature issue and will follow up with a PR that implements `admin.groups` — an optional ordered array of group labels that `groupNavItems` sorts against. Backward compatible: groups not in the array keep the existing first-encounter order. Keeps the design sketch from this discussion but scoped to group-level ordering only.
